cranelift-codegen/src/isa/x86/abi.rs

```diff
@@ -752,6 +838,49 @@ fn insert_common_prologue(
         }
     }
 
+    if csrs.iter(FPR).len() != 0 {
+        let mut fpr_offset = 0;
+
+        // `stack_store` is not directly encodable in x86_64 at the moment, so we'll need a base
+        // address. We are well after postopt could run, so load the CSR region base once here,
+        // instead of hoping that the addr/store will be combined later.
+        let stack_addr = pos.ins().stack_addr(types::I64, *csr_slot, 0);
+
+        // At this point we won't be using volatile registers for anything except for return
+        // At this point we have not saved any CSRs yet, and arguments are all in registers. So
+        // pick a location for `stack_addr` that is neither non-volatile nor an argument: RAX will
+        // do.
+        pos.func.locations[stack_addr] = ir::ValueLoc::Reg(RU::rax as u16);
+
+        for reg in csrs.iter(FPR) {
+            // Append param to entry Block
+            let csr_arg = pos.func.dfg.append_block_param(block, types::F64);
+
+            // Assign it a location
+            pos.func.locations[csr_arg] = ir::ValueLoc::Reg(reg);
+
+            let reg_store_inst =
+                pos.ins()
+                    .store(ir::MemFlags::trusted(), csr_arg, stack_addr, fpr_offset);
+            fpr_offset += types::F64X2.bytes() as i32;
```

Comment:
Wait -- shouldn't `fpr_offset` be initialized to the offset chosen for the stack slot earlier? Right now it is initialized to 0. (Or alternatively but equivalently use the stack slot offset as an immediate for the `stack_addr` instruction?) Otherwise we are always reading xmm6-15 from stack frame offsets `0..`. Is this the source of the overlapping stack slot bug?

Reply:
same as above, either this _is_ or is closely related to the overlapping stack slot issue. Either this will change or I'll update (probably with a comment describing CSR region layout) when it's fixed.

Reply:
(a nice benefit of picking the end of the CSR region is that it is defined to be aligned by ABI requirements, so we don't need to figure out if the last GPR save left the stack aligned or not)